src/database/statement.ts

```diff
@@ -9,11 +9,18 @@ export default class Statement {
     });
   }
 
-  static delimName(name: string) {
-    if (name.startsWith(`"`) && name.endsWith(`"`)) return name;
-    if (name.includes(` `)) return `"${name}"`;
+  /**
+   * 
+   * @param name Value which should be normalised
+   * @param fromUser If the value is true, then we likely need to normalise. Items from the database are usually normalised already
+   * @returns 
+   */
+  static delimName(name: string, fromUser = false) {
+    if (fromUser && name.startsWith(`"`) && name.endsWith(`"`)) return name;
+    if (fromUser === false && name.includes(` `)) return `"${name}"`;
     if (name.length <= 10) return name.toUpperCase();
-    else return `"${name}"`;
+    else if (fromUser) return `"${name}"`;
+    else return name;
   }
```

Comment:
I think this still needs tweaking.  This is not complete, but gives an idea of what I think is needed.
``` typescript
    if (fromUser) { // The name was input by the user
      // If already delimited, return it as-is
      if (name.startsWith(`"`) && name.endsWith(`"`)) return name;
      // If the value contains a space or decimal it needs to be delimited
      if (name.includes(` `) || name.includes(`.`)) return `"${name}"`;
      // Otherwise, fold to uppercase.  The user should have explicitly delimited if that was their intention.
      return name.toUpperCase();
    } else { // The name came from a catalog file query
      // If the name contains characters other than the valid variants, uppercase, digits, or underscores, it must be delimited
      // Note: looks like there is some good logic in vscode-ibmi\src\api\IBMi.ts\parserMemberPath that has a regex that is variant-character aware
      if (name.match(`[^A-Z0-9_@#$]`)) return `"${name}"`;
      // The name should already be uppercase coming from the catalog so just return it
      return name;
    }
```

Reply:
Your comments made it much easier for me to understand, thanks. I went ahead and used this as a base for the change and also implemented `validQsysName` which uses the variant characters from the connection to validate it. A new commit is in. Let me know what you think.